File: lua/test/book_entity_test.lua
-- Book entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("stephen-king_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("BookEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Book(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = book_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "book." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set STEPHENKING_TEST_BOOK_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local book_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.book")))
    local book_ref01_data = nil
    if #book_ref01_data_raw > 0 then
      book_ref01_data = helpers.to_map(book_ref01_data_raw[1][2])
    end

    -- LIST
    local book_ref01_ent = client:Book(nil)
    local book_ref01_match = {}

    local book_ref01_list_result, err = book_ref01_ent:list(book_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(book_ref01_list_result)

    -- LOAD
    local book_ref01_match_dt0 = {
      id = book_ref01_data["id"],
    }
    local book_ref01_data_dt0_loaded, err = book_ref01_ent:load(book_ref01_match_dt0, nil)
    assert.is_nil(err)
    local book_ref01_data_dt0_load_result = helpers.to_map(book_ref01_data_dt0_loaded)
    assert.is_not_nil(book_ref01_data_dt0_load_result)
    assert.are.equal(book_ref01_data_dt0_load_result["id"], book_ref01_data["id"])

  end)
end)

function book_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/book/BookTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read book test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "book01", "book02", "book03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("STEPHENKING_TEST_BOOK_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["STEPHENKING_TEST_BOOK_ENTID"] = idmap,
    ["STEPHENKING_TEST_LIVE"] = "FALSE",
    ["STEPHENKING_TEST_EXPLAIN"] = "FALSE",
    ["STEPHENKING_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["STEPHENKING_TEST_BOOK_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["STEPHENKING_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["STEPHENKING_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["STEPHENKING_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["STEPHENKING_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: ts/test/entity/book/BookEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { StephenKingSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('BookEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when STEPHENKING_TEST_LIVE=TRUE.
  afterEach(liveDelay('STEPHENKING_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = StephenKingSDK.test()
    const ent = testsdk.Book()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.STEPHEN_KING_TEST_LIVE
    for (const op of ['list', 'load']) {
      if (maybeSkipControl(t, 'entityOp', 'book.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set STEPHEN_KING_TEST_BOOK_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let book_ref01_data = Object.values(setup.data.existing.book)[0] as any

    // LIST
    const book_ref01_ent = client.Book()
    const book_ref01_match: any = {}

    const book_ref01_list = await book_ref01_ent.list(book_ref01_match)


    // LOAD
    const book_ref01_match_dt0: any = {}
    book_ref01_match_dt0.id = book_ref01_data.id
    const book_ref01_data_dt0 = await book_ref01_ent.load(book_ref01_match_dt0)
    assert(book_ref01_data_dt0.id === book_ref01_data.id)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/book/BookTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = StephenKingSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['book01','book02','book03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['STEPHEN_KING_TEST_BOOK_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'STEPHEN_KING_TEST_BOOK_ENTID': idmap,
    'STEPHEN_KING_TEST_LIVE': 'FALSE',
    'STEPHEN_KING_TEST_EXPLAIN': 'FALSE',
    'STEPHEN_KING_APIKEY': 'NONE',
  })

  idmap = env['STEPHEN_KING_TEST_BOOK_ENTID']

  const live = 'TRUE' === env.STEPHEN_KING_TEST_LIVE

  if (live) {
    client = new StephenKingSDK(merge([
      {
        apikey: env.STEPHEN_KING_APIKEY,
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.STEPHEN_KING_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  
